src/Evolution/DiscontinuousGalerkin/Actions/NormalVectorQuantities.hpp

```diff
@@ -0,0 +1,176 @@
+
+namespace evolution::dg::Actions::detail {
+CREATE_HAS_TYPE_ALIAS(inverse_spatial_metric_tag)
+CREATE_HAS_TYPE_ALIAS_V(inverse_spatial_metric_tag)
+
+template <bool HasInverseSpatialMetricTag = false>
+struct inverse_spatial_metric_tag_impl {
+  template <typename System>
+  using f = tmpl::list<>;
+};
+
+template <>
+struct inverse_spatial_metric_tag_impl<true> {
+  template <typename System>
+  using f = tmpl::list<typename System::inverse_spatial_metric_tag>;
+};
+
+template <typename System>
+using inverse_spatial_metric_tag = typename inverse_spatial_metric_tag_impl<
+    has_inverse_spatial_metric_tag_v<System>>::template f<System>;
+
+struct OneOverNormalVectorMagnitude : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim>
+struct NormalVector : db::SimpleTag {
+  using type = tnsr::I<DataVector, Dim, Frame::Inertial>;
+};
+
+/*!
+ * \brief Computes the normal covector, magnitude of the unnormalized normal
+ * covector, and in curved spacetimes the normal vector.
+ */
+template <typename System, size_t Dim, typename FieldsOnFaceTags>
+void unit_normal_vector_and_covector(
+    const gsl::not_null<Scalar<DataVector>*> face_normal_magnitude,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        unit_normal_covector,
+    const gsl::not_null<Variables<FieldsOnFaceTags>*> fields_on_face,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unnormalized_normal_covector) noexcept {
+  if constexpr (has_inverse_spatial_metric_tag_v<System>) {
+    using inverse_spatial_metric_tag =
+        typename System::inverse_spatial_metric_tag;
+    auto& normal_vector = get<NormalVector<Dim>>(*fields_on_face);
+    const auto& inverse_spatial_metric =
+        get<inverse_spatial_metric_tag>(*fields_on_face);
+    // Compute unnormalized normal vector
+    for (size_t i = 0; i < Dim; ++i) {
+      normal_vector.get(i) = inverse_spatial_metric.get(i, 0) *
+                             get<0>(unnormalized_normal_covector);
+      for (size_t j = 1; j < Dim; ++j) {
+        normal_vector.get(i) += inverse_spatial_metric.get(i, j) *
+                                unnormalized_normal_covector.get(j);
+      }
+    }
+    // Perform normalization
+    dot_product(face_normal_magnitude, normal_vector,
+                unnormalized_normal_covector);
+    get(*face_normal_magnitude) = sqrt(get(*face_normal_magnitude));
+    auto& one_over_normal_vector_magnitude =
+        get<OneOverNormalVectorMagnitude>(*fields_on_face);
+    get(one_over_normal_vector_magnitude) = 1.0 / get(*face_normal_magnitude);
+    for (size_t i = 0; i < Dim; ++i) {
+      unit_normal_covector->get(i) = unnormalized_normal_covector.get(i) *
+                                     get(one_over_normal_vector_magnitude);
+      normal_vector.get(i) *= get(one_over_normal_vector_magnitude);
+    }
+  } else {
+    magnitude(face_normal_magnitude, unnormalized_normal_covector);
```

Comment:
There are a few duplicated lines here. Could you avoid repeating them somehow, maybe with an if constexpr?

Reply:
I don't really know how to do that without adding a second `if constexpr` block that does the same checks, and personally I would find that a lot harder to read.  Really the only lines that would be common are:
```cpp
auto& one_over_normal_vector_magnitude =
        get<OneOverNormalVectorMagnitude>(*fields_on_face);
    get(one_over_normal_vector_magnitude) = 1.0 / get(*face_normal_magnitude);
```
and add 1 blank line plus 3 for the `if-else` seems worse than having that statement twice

Reply:
Fair enough